tests/org.eclipse.ui.tests.navigator/src/org/eclipse/ui/tests/navigator/GoBackForwardsTest.java

```diff
@@ -0,0 +1,252 @@
+import org.eclipse.ui.IEditorInput;
+import org.eclipse.ui.INavigationLocation;
+import org.eclipse.ui.IWorkbenchPage;
+import org.eclipse.ui.IWorkbenchWindow;
+import org.eclipse.ui.PartInitException;
+import org.eclipse.ui.PlatformUI;
+import org.eclipse.ui.internal.NavigationHistoryAction;
+import org.eclipse.ui.intro.IIntroPart;
+import org.eclipse.ui.part.FileEditorInput;
+import org.eclipse.ui.tests.harness.util.EditorTestHelper;
+import org.eclipse.ui.tests.harness.util.FileTool;
+import org.eclipse.ui.tests.harness.util.FileUtil;
+import org.eclipse.ui.tests.harness.util.UITestCase;
+import org.eclipse.ui.texteditor.AbstractTextEditor;
+import org.eclipse.ui.texteditor.TextSelectionNavigationLocation;
+import org.junit.Assert;
+import org.junit.Test;
+
+/**
+ * @since 3.3
+ *
+ */
+public class GoBackForwardsTest extends UITestCase {
+
+	private static final String TEST_NAME = "GoBackForwardsTest";
+
+	public GoBackForwardsTest() {
+		super(TEST_NAME);
+	}
+
+	private static final String PROJECT_NAME = "GoBackForwardsTestProject";
+	private static final String FILE_NAME = "GoBackForwardsTestFile.java";
+	private static final String FILE_CONTENTS = "public class GoBackForwardsTestFile {\n"
+			+ "    public static void main(String[] args) {\n" + "        System.out.println(\"Hello world!\");\n"
+			+ "    }\n" + "}";
+	private static final String JAVA_EDITOR_ID = "org.eclipse.jdt.ui.CompilationUnitEditor";
+	private static final String TEXT_EDITOR_ID = "org.eclipse.ui.genericeditor.GenericEditor";
+	private static final String SELECTION_STRING = "Selection<offset: 10, length: 5>";
+
+	private IProject project;
+	private IFile file;
+
+	@Override
+	public void doSetUp() {
+		try {
+			project = FileUtil.createProject(PROJECT_NAME);
+			file = FileUtil.createFile(FILE_NAME, project);
+			StringBuilder stringBuilder = new StringBuilder();
+			stringBuilder.append(FILE_CONTENTS);
+			FileTool.writeFromBuilder(file.getLocation().toOSString(), stringBuilder);
+		} catch (CoreException e) {
+			fail("Should not throw an exception");
+		} catch (IOException e) {
+			fail("Should not throw an exception");
+		}
+
+	}
+
+	@Test
+	public void testNavigationHistoryNavigation() {
+		IIntroPart introPart = PlatformUI.getWorkbench().getIntroManager().getIntro();
+		PlatformUI.getWorkbench().getIntroManager().closeIntro(introPart);
+
+		Condition javaEditorNoSelection = currentNavigationHistoryLocationCondition(JAVA_EDITOR_ID, false);
+		Condition javaEditorSelection = currentNavigationHistoryLocationCondition(JAVA_EDITOR_ID, true);
+		Condition textEditorNoSelection = currentNavigationHistoryLocationCondition(TEXT_EDITOR_ID, false);
+		Condition textEditorSelection = currentNavigationHistoryLocationCondition(TEXT_EDITOR_ID, true);
+
+		FileEditorInput editorInput = new FileEditorInput(file);
+
+		openJavaEditor(editorInput);
+
+		if (!processEventsUntil(javaEditorNoSelection, 30000)) {
+			fail("Timeout during navigation.");
+		}
+		;
```

Comment:
Just one small nit.  Could you please get rid of all these lines with just a semicolon on them?